Docs release 1.10.0 csi 4365 host define (#517)
IBM/ibm-block-csi-driver: docs/content/configuration/configuring_hostdefiner.md

# Configuring the host definer

Some of the parameters within the HostDefiner custom resource are configurable. Use this information to help decide whether the parameters for your storage system need to be updated.

For more information about using the host definer, see [Using dynamic host definition](../using/using_hostdefinition.md).
    
|Field|Description|
|---------|--------|

> |`prefix`|Adds a prefix to the hosts defined by the host definer.|

Reviewer: We should add a note of the max chars and chars types supported for host name in the storage (currently only for SVC).

Author: @roysahar  @matancarmeli7  what are the max characters and allowed character types here?
regarding SVC limitation - this is documented in limitations

Reviewer 2: I will check and update this

Reviewer 2: this is Carlos's answer:
"63 characters, must start alpha, characterset is printable characters (not sure what special characters are allowed)"

Author: @erantzabari  how do we want to define this in the guide (if at all)
@oriyarde 
note that in https://www.ibm.com/docs/en/flashsystem-9x00/8.5.x?topic=commands-host host_name only states that it needs to be alphanumeric. is this the relevant place to look?

Reviewer 2: @oriyarde I don't know, I am not sure.
So if we allow the user to put 20 chars as a prefix, he won't be able to have a host with more than 42 chars in his name?

Reviewer 3: if we don't have a clever "final name" logic, and the storage command fails for more than 63 chats, then it seems like yes

Reviewer 3: could we write 20 chars for now, until we have a final decision? we cannot afford to miss this

Reviewer 4: We can add a comment that prefix and host_name are bounded to the limitation of the storage system and the rules from the storage system applies also for the prefix.

Author: updated per eran's input. @oriyarde please approve
https://github.com/IBM/ibm-block-csi-driver/blob/83cbf8101de7752cd6afa57f8ee1fbd70bdd1278/docs/content/configuration/configuring_hostdefiner.md

Reviewer 3: fine, but this means that if the user wants a combination of prefix, separator and worker name which is longer than want the storage allows, we could never add a clever hashing mechanism (as we did for volumes and snapshots)

Reviewer 4: @rivkap , we can add comment in the documentation on top of the comment that we bound to the limitation of the storage, that the prefix is part of the host name (so the length is combined to both the prefix and the name given).

Reviewer 4: @oriyarde , I wouldn't go to hashing solution. We added it to DS8K due to very short volume name - 63 chars are good enough for prefix and name.

Author: thanks. updated per @erantzabari 's input

Reviewer 3: @erantzabari
1. we added volume name hashing to all storage systems
2. a valid k8s hostname can be up to 253 chars long, even without a prefix
3. in the future, we might add host definer support for stricter storage systems

we can always remove the prefix length limitation, but we cannot add retroactively. I don't see why not being on the safe side...

Reviewer 4: We added the limitation of the length according to the storage system (not specific number).
Do you suggest something else or you good with that?

Reviewer 3: I suggest something else: [limit the prefix length](https://github.com/IBM/ibm-block-csi-driver/pull/517#discussion_r921226599) (as we did with volume prefix), to allow adding hashing in the future, otherwise a long prefix would render the hashing useless (due to collisions)

|`connectivity`|Selects the connectivity type for the host ports.<br>Possible input values are:<br>- `nvme` for use with NVME over Fibre Channel connectivity<br>- `fc` for use with Fibre Channel over SCSI connectivity<br>- `iscsi` for use with iSCSI connectivity<br><br>By default, this field is blank and the host definer selects the first of available connectivity types on the node, according to the following hierarchy: NVMe, FC, iSCSI.|

> |`allowDelete`|Defines whether the host definer is allowed to delete host definitions.<br>Input values are `true` or `false`.<br>The default value is `true`.|

Reviewer: Please add description that not just hostdefinition will be deleted but the host object from the storage.

Author: updated

Reviewer 3: >is allowed to delete host definitions

this means definitions on storage. an "object" is just a general term that includes a "definition".
to me it reads as if the same thing is written twice.
AFAIK we **never** refer to the k8s hostdefinition with a space between "host" and "definition".
@matancarmeli7 ?

Author: updated in meeting with @matancarmeli7 and @roysahar

Reviewer 3: IMO it is more important for the user to know if the definer is allowed to delete definitions on storage rather than to delete its internal `HostDefinition` k8s objects

Author: @oriyarde We decided to word this as follows based off of the following:
1. When describing the feature in [Using dynamic host definition](https://github.com/IBM/ibm-block-csi-driver/blob/docs_release-1.10.0/docs/content/using/using_hostdefinition.md) we state there that this is happening on the storage.
2. When talking about [configuration for PVC](https://github.com/IBM/ibm-block-csi-driver/blob/docs_release-1.10.0/docs/content/configuration/creating_pvc.md) we don't mention storage side each time.

@roysahar  @matancarmeli7  please feel free to expand on this or correct anything that I missed / misunderstood based on our conversation the other day. thanks

Reviewer 3: 1. true, though we didn't explicitly write about the definer **deleting** definitions on storage
2. a PVC is managed by the user, a `HostDefinition` is managed by the definer. in essence, the user shouldn't care much about the existence of the k8s `HostDefinition` object (unless they want to troubleshoot/verify actions from the k8s side)

the reason the `allowDelete` parameter exists is in case the user relies on the storage definitions for non-CSI usage, and manually deleting the k8s `HostDefinition` object doesn't make the definer delete the storage definitions, so what matters is the deletion on storage, not in k8s

Author: so should I just add "on the storage system" to the end of the sentence?

Reviewer 3: not exactly, because we also use a different casing convention for each side. I suggest:
```
Defines whether the host definer is allowed to delete host definitions on the storage system.
```

Author: updated and can be found [here](https://github.com/IBM/ibm-block-csi-driver/blob/docs_release-1.10.0/docs/content/configuration/configuring_hostdefiner.md)

Reviewer 3: >we also use a different casing convention for each side

so `HostDefinitions on the storage system` should be `host definitions on the storage system`

Author: From my conversation with @roysahar  and @matancarmeli7  we were talking about the HostDefinition file itself. is this no longer the case when talking about on the storage side?

Reviewer 3: AFAIK it is (and it has always been) `HostDefinition` for the k8s side, and "host definition" for the storage side

Author: I updated. @roysahar  please confirm that this shoudl be "host definitions on the storage system" and _not_ "HostDefinitions on storage system"

Reviewer 2: confirmed

Author: this is showing on https://github.com/IBM/ibm-block-csi-driver/pull/532/files#diff-b882cc376223979bdac6e43eb33e525baefc16e5d726e0cc879ca309206675f4

> |`dynamicNodeLabeling`|Defines whether the nodes that run the CSI node pod are dynamically labeled or if the user must create the `hostdefiner.block.csi.ibm.com/managed-by=true` label on each relevant node. This label tells the host definer which nodes to manage their host definition on the storage side.<br>Input values are `true` or `false`.<br>The default value is `false`, where the user must create each label.|

Reviewer: Should be label - `hostdefiner.block.csi.ibm.com/manage-node`

Author: updated in all instances

Reviewer: Thanks.

> |`dynamicNodeLabeling`|Defines whether the nodes that run the CSI node pod are dynamically labeled or if the user must create the `hostdefiner.block.csi.ibm.com/managed-by=true` label on each relevant node. This label tells the host definer which nodes to manage their host definition on the storage side.<br>Input values are `true` or `false`.<br>The default value is `false`, where the user must create each label.|

Reviewer: If false - "where the user will have to create manually this label on every node he wants to be managed by host definer for automatic host define on the storage"

Author: updated

Reviewer: Thanks.

> |`dynamicNodeLabeling`|Defines whether the nodes that run the CSI node pod are dynamically labeled or if the user must create the `hostdefiner.block.csi.ibm.com/managed-by=true` label on each relevant node. This label tells the host definer which nodes to manage their host definition on the storage side.<br>Input values are `true` or `false`.<br>The default value is `false`, where the user must create each label.|

Reviewer: Also need to add note that these labels won't be deleted when hostdefiner will be removed, and explain on which situation the label will be deleted automatically when the value set to "true".
Also there should be a note on a case which the label/s won't succeed to be deleted see - https://jira.xiv.ibm.com/browse/CSI-4446

Author: is CSI-4446 a known issue or just part of the feature currently?

Reviewer 2: @roysahar I asked @erantzabari about what to write about the labels won't be deleted when hostdefiner will be removed, and he said that we need to add this note at the uninstall process and it's enough
![image](https://user-images.githubusercontent.com/45543087/178969728-f24fe0ab-393d-416d-a883-85f0d9eb2cc8.png)

Author: "and explain on which situation the label will be deleted automatically when the value set to "true"."

what is the specific situation that i should say ?

Author: Currently I already updated 
![image](https://user-images.githubusercontent.com/54570654/178971513-e7bbb029-f155-409c-aa72-954927278d51.png)

should this be removed? should I do any other updates for this section or per Matan this is ok as is?

@matancarmeli7  @roysahar  thanks

Author: in my meeting with @roysahar  and @matancarmeli7  We removed the note in the end that was added to the parameter.
Needs to be decided by @erantzabari : should https://jira.xiv.ibm.com/browse/CSI-4446 be listed as a known issue / limitation in the guide.

Reviewer 3: @rivkap please avoid using internal links in public. a ticket number should be enough

Author: @roysahar  can we close this thread?

Author: done.

> |dynamicNodeLabeling|Defines whether the nodes that run our CSI node pod are dynamically labelled or if the user must create the `hostdefiner.block.csi.ibm.com/managed-by=true` label on each relevant node. In addition, this label allows the host definer to find all nodes with this label and manages their storage definition.<br>Input values are `true` or `false`.<br>The default value is `false`, where the user must create each label.|

Reviewer 3: ```suggestion
|dynamicNodeLabeling|Defines whether the nodes that run the CSI node pod are dynamically labelled or if the user must create the `hostdefiner.block.csi.ibm.com/managed-by=true` label on each relevant node. This label tells the host definer which nodes to manage their host definition on storage.<br>Input values are `true` or `false`.<br>The default value is `false`, where the user must create each label.|
```
AFAIK we never us "our" in official docs to describe the product we develop.
`In addition` makes it sound like a separate thing, when it's actually an elaboration of the first sentence.
`this label allows the host definer to find all nodes with this label` sounds too obvious and is not the main focus.
`storage definition` might sound like a definition of a storage, that is not a host definition

Author: done

> |`dynamicNodeLabeling`|Defines whether the nodes that run the CSI node pod are dynamically labeled or if the user must create the `hostdefiner.block.csi.ibm.com/manage-node=true` label on each relevant node. This label tells the host definer which nodes to manage their host definition on the storage side.<br>Input values are `true` or `false`.<br>The default value is `false`, where the user must manually create this label on every node to be managed by the host definer for automatic host definition on the storage.<br><br>**Note:** When the label is set to `false`, these labels are not deleted during host definer uninstall.|

Reviewer 3: >**Note:** When the label is set to `false`, these labels are not deleted during host definer uninstall.

> When **the** label

which label? AFAIK `dynamicNodeLabeling` is not a label. I'm confused

Author: it was a typo - it should be "When the parameter is set to `false` these labels are not deleted during host definer uninstall."

is this OK?

Reviewer 3: yes, thanks

Author: updated in source file. thanks
